# Проект "Изучение игровой индустрии"

- Автор: Денисова Валерия
- Дата: 09.01.2026

### Цели и задачи проекта

В рамках данного проекта мы познакомимся с датасетом `new_games.csv`, содержащим данные о развитии игровой индустрии, проведем предобработку данных, сделаем фильтрацию - оставим данные только за 2000-2013 гг., которые нужны заказачику, а также проведем категоризацию данных и сделаем выводы.

### Описание данных

Данные `new_games.csv` содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:
- `Name` — название игры.
- `Platform` — название платформы.
- `Year of Release` — год выпуска игры.
- `Genre` — жанр игры.
- `NA sales` — продажи в Северной Америке (в миллионах проданных копий).
- `EU sales` — продажи в Европе (в миллионах проданных копий).
- `JP sales` — продажи в Японии (в миллионах проданных копий).
- `Other sales` — продажи в других странах (в миллионах проданных копий).
- `Critic Score` — оценка критиков (от 0 до 100).
- `User Score` — оценка пользователей (от 0 до 10).
- `Rating` — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.

### Содержимое проекта
1. Загрузка данных и знакомство с ними
2. Проверка ошибок в данных и их предобработка
3. Фильтрация данных
4. Категоризация данных

## 1. Загрузка данных и знакомство с ними

- Загрузим необходимые библиотеки Python и данные датасета `new_games.csv`.


In [106]:
# импортируем библиотеку pandas для работы с данными
import pandas as pd

In [107]:
# выгрузим данные из файла 'new_games.csv' в датафрейм games
games = pd.read_csv('new_games.csv')

- Познакомимся с данными: выведем первые и последние строки датасета и результат метода `info()`.


In [108]:
# выведем общую информацию о датафрейме
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [109]:
# сохраним в переменную количсетво строк в исходном датасете
initial_len = games.shape[0]
initial_len

16956

In [110]:
# выведем первые 5 строк датафрейма
games.head(5) 

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


In [111]:
# выведем последние 5 строк датафрейма
games.tail()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
16951,Samurai Warriors: Sanada Maru,PS3,2016.0,Action,0.00,0.0,0.01,0.0,NaN,NaN,NaN
16952,LMA Manager 2007,X360,2006.0,Sports,0.00,0.01,0.0,0.0,NaN,NaN,NaN
16953,Haitaka no Psychedelica,PSV,2016.0,Adventure,0.00,0.0,0.01,0.0,NaN,NaN,NaN
16954,Spirits & Spells,GBA,2003.0,Platform,0.01,0.0,0.0,0.0,NaN,NaN,NaN
16955,Winning Post 8 2016,PSV,2016.0,Simulation,0.00,0.0,0.01,0.0,NaN,NaN,NaN


Датасет `new_games.csv` содержит `16956 строчек и 11 столбцов`. Данные соответствуют описанию. Используются следующие типы данных:
- `object` (7): для столбцов Name, Platform, Genre, Rating - этот тип данных вполне подходит, так как эти столбцы содержат текстовые данные. Для столбцов EU sales, JP sales, User Score тип данных `object` не является подходящим, т.к. они содержат число, поэтому лучше преобразовать тип данных к числовому `float`, чтобы была возможность выполнять операции с числами.
- `float64` (4): полям  Year of Release, NA sales, Other sales, Critic Score подходит тип `float64`, т.к. они содержат числовые значения.

Названия столбцов прописаны в неудобном формате - в разном регистре и с пробелами, поэтому стоит привести их к единому стилю для удобства. Также можно заметить, что больше всего пропусков содержится в столбцах с оценками критиков, пользователей и рейтингом. Возможно это связано с тем, что это все субъективные оценки, и не было найдено какого-то конкретного точного значения, либо с тем, что игра вышла недавно и еще просто не успела собрать достаточное количество оценок.

---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

- Выведем на экран названия всех столбцов датафрейма и проверим их стиль написания.
- Приведем все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

In [112]:
# выведем названия всех столбцов датафрейма
games.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [113]:
# заменим в названиях столбцов пробел на подчеркивание и приведем все названия к нижнему регистру
names = []
for name in games.columns:
    name = name.replace(' ', '_').lower()
    names.append(name)
games.columns = names

Мы привели названия столбцов датафрейма к более удобному формату для дальнейшей работы с ними. 

### 2.2. Типы данных

Ранее мы выяснили, что в полях `eu_sales, jp_sales, user_score` стоит изменить тип данных на `float`,  в `year_of_release` на `int`.

In [114]:
''' преобразуем данные в столбцах eu_sales, jp_sales, user_score к типу float, используя параметр downcast, чтобы подобрать самый оптимальный тип 
данных, и параметр coerce, чтобы нечисловые значения заменились на Nan '''
games['eu_sales'] = pd.to_numeric(games['eu_sales'], errors = 'coerce', downcast = 'float')
games['jp_sales'] = pd.to_numeric(games['jp_sales'], errors = 'coerce', downcast = 'float')
games['user_score'] = pd.to_numeric(games['user_score'], errors = 'coerce', downcast = 'float')

Также изменим тип данных в столбцах `na_sales`, `other_sales` и `critic_score` с float64 на более оптимальный.

In [115]:
# с помощью параметра downcast подбираем самый оптимальный тип данных
games['na_sales'] = pd.to_numeric(games['na_sales'], downcast = 'float')
games['other_sales'] = pd.to_numeric(games['other_sales'], downcast = 'float')
games['critic_score'] = pd.to_numeric(games['critic_score'], downcast = 'float')

In [116]:
# проверим, что у нас получилось
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float32
 5   eu_sales         16950 non-null  float32
 6   jp_sales         16952 non-null  float32
 7   other_sales      16956 non-null  float32
 8   critic_score     8242 non-null   float32
 9   user_score       7688 non-null   float32
 10  rating           10085 non-null  object 
dtypes: float32(6), float64(1), object(4)
memory usage: 1.0+ MB


Заметим, что размер занимаемой памяти значительно уменьшился.

Для оптимизации работы с данными в датафрейме были сделаны следующие изменения типов данных:
- `eu_sales`, `jp_sales`, `user_score`: тип изменен с `object` на `float32`
- `year_of_release`: тип изменен с `float64` на `int16`
- `na_sales`, `other_sales`, `critic_score`: тип изменен с `float64` на `float32`

### 2.3. Наличие пропусков в данных

- Посчитаем количество пропусков в каждом столбце в абсолютных и относительных значениях.


In [117]:
# посчитаем количество пропусков в каждом столбце
games.isna().sum()

name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8714
user_score         9268
rating             6871
dtype: int64

In [118]:
# посчитаем процент пропущенных значений от всех в каждом столбце
games.isna().mean() * 100 

name                0.011795
platform            0.000000
year_of_release     1.621845
genre               0.011795
na_sales            0.000000
eu_sales            0.035386
jp_sales            0.023590
other_sales         0.000000
critic_score       51.391838
user_score         54.659118
rating             40.522529
dtype: float64

В столбце `name` содержится 2 пропуска (0, 011795%). Это мало, можно проигнорировать, так как для исследования нам важно не столько название игры, сколько её жанр и количество продаж. Но нужно проверить, есть ли в строках с пропущенным названием жанр.

In [119]:
# выведем строки, где название отсутствует
games[games['name'].isna()]

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
661,NaN,GEN,1993.0,NaN,1.78,0.53,0.00,0.08,NaN,NaN,NaN
14439,NaN,GEN,1993.0,NaN,0.00,0.00,0.03,0.00,NaN,NaN,NaN


Заметим, что в строках с пропущенным названием отсутствует и данные в поле `genre`. В таком случае, две эти строки можно удалить, т.к. они не дадут нам никакой полезной информации при анализе.

В столбце `year_of_release` были пропуски, которые мы ранее обработали и заменили на -1, т.к. год точно не может быть равен -1. Посмотрим, много ли таких строк.

In [120]:
# посмотрим в скольких строках был пропущен год
games[games['year_of_release'].isna()].count()

name               275
platform           275
year_of_release      0
genre              275
na_sales           275
eu_sales           275
jp_sales           275
other_sales        275
critic_score       157
user_score         130
rating             184
dtype: int64

Мы не будем удалять строки с пропущенным значением года, т.к. их 275, и данные из них все равно могут пригодиться нам для анализа продаж в разрезе жанров игр. Но в дальнейшем при фильтрации данных за 2000-2013 год эти данные все равно уберутся и не исказят результаты анализа.

В столбцах `eu_sales` и `jp_sales` содержится 6 и 4 пропусков соответственно. Они могли появиться из-за технической ошибки при сборе данных, либо из-за отсутсвия продаж в данных странах. Количество пропусков небольшое (примерно 0.035% и 0.024%), поэтому можно заменить их на среднее количество продаж для такой платформы и года выхода игры.

Самое большое количество пропусков в столбцах `critic_score` (8714), `user_score` (9268), `rating` (6871) - примерно 51.4%, 54.7, 40.5% соответственно, что является очень большим количеством. Как мы ранее предположили, это может быть связано с тем, что это все субъективные оценки, и не было найдено какого-то конкретного точного значения, либо с тем, что игра вышла недавно и еще просто не успела собрать достаточное количество оценок. В данном случае мы не будем удалять строки, т.к. они составляют очень большой процент от данных и остальные данные могут быть нам важны. Заменять на среднее значение мы тоже не будем, т.к. это субъективные оценки, и слишком мало данных, на основе которых можно рассчитать среднее, это может исказить результаты анализа. В столбцах `critic_score` и `user_score` заменим пропуски на значение -1, чтобы сохранить числовой тип данных и т.к. оценки не могут быть отрицательными. В столбце `rating` заменим пропуски на текстовый комментарий "Нет данных".

Удалим строки, где нет информации о названии и жанре.

In [121]:
# найдем индексы строк с пропущенным названием
index = games[games['name'].isna()].index
# удалим строки с этими индексами
games = games.drop(index)

Заменим пропуски в столбцах `eu_sales` и `jp_sales` на среднее значение для игр с той же платформы и того же года выпуска.

In [122]:
# создадим функцию, которая будет возвращать среднее значение eu_sales для группы игр с той же платформы и того же года выпуска, если значение пропущено 
def change_eu(row):
    if row.isna()['eu_sales']:
        group = games[(games['platform'] == row['platform']) & (games['year_of_release'] == row['year_of_release'])]
        return group['eu_sales'].mean()
    else:
        return row['eu_sales']
# применим функцию ко всем строкам датафрейма
games['eu_sales'] = games.apply(change_eu, axis = 1)

In [123]:
# создадим такую же функцию для jp_sales 
def change_jp(row):
    if row.isna()['jp_sales']:
        group = games[(games['platform'] == row['platform']) & (games['year_of_release'] == row['year_of_release'])]
        return group['jp_sales'].mean()
    else:
        return row['jp_sales']
# применим функцию ко всем строкам датафрейма
games['jp_sales'] = games.apply(change_jp, axis = 1)

Заменим пропуски в столбцах `critic_score`, `user_score` на -1, а в `rating` на "Нет данных".

In [124]:
# используем метод fillna для заполнения пропусков
games['critic_score'] = games['critic_score'].fillna(-1)
games['user_score'] = games['user_score'].fillna(-1)
games['rating'] = games['rating'].fillna('Нет данных')

### 2.4. Явные и неявные дубликаты в данных

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

Изучим уникальные значения в поле `genre`.

In [125]:
# выведем уникальные значения поля genre
games['genre'].unique()

array(['Sports', 'Platform', 'Racing', 'Role-Playing', 'Puzzle', 'Misc',
       'Shooter', 'Simulation', 'Action', 'Fighting', 'Adventure',
       'Strategy', 'MISC', 'ROLE-PLAYING', 'RACING', 'ACTION', 'SHOOTER',
       'FIGHTING', 'SPORTS', 'PLATFORM', 'ADVENTURE', 'SIMULATION',
       'PUZZLE', 'STRATEGY'], dtype=object)

Заметим, что значения повторяются из-за разного регистра. Приведем все названия к нижнему регистру.

In [126]:
# приводим все названия к нижнему регистру
games['genre'] = games['genre'].str.lower()
# выводим уникальные значения
games['genre'].unique()

array(['sports', 'platform', 'racing', 'role-playing', 'puzzle', 'misc',
       'shooter', 'simulation', 'action', 'fighting', 'adventure',
       'strategy'], dtype=object)

Проверим уникальные значения `year_of_release`.

In [127]:
# выведем уникальные года в порядке возрастания
games['year_of_release'].sort_values().unique()

array([1980., 1981., 1982., 1983., 1984., 1985., 1986., 1987., 1988.,
       1989., 1990., 1991., 1992., 1993., 1994., 1995., 1996., 1997.,
       1998., 1999., 2000., 2001., 2002., 2003., 2004., 2005., 2006.,
       2007., 2008., 2009., 2010., 2011., 2012., 2013., 2014., 2015.,
       2016.,   nan])

В годах отсутсвуют дубликаты.

Далее проверим уникальные значения `platform`.

In [128]:
# выведем уникальные платформы
games['platform'].unique()

array(['Wii', 'NES', 'GB', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       'PS4', '3DS', 'N64', 'PS', 'XB', 'PC', '2600', 'PSP', 'XOne',
       'WiiU', 'GC', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

В платформах отсутствуют дубликаты, проверим `rating`:

In [129]:
games['rating'].unique()

array(['E', 'Нет данных', 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP'],
      dtype=object)

K-A - это устаревшее обозначение рейтинга E. Система ESRB заменила "Kids to Adults" (K-A) на "Everyone" (E) в 1998 году. Заменим "K-A" на "E". 

In [130]:
# с помощью метода replace заменим 'K-A' на 'E'
games['rating'] = games['rating'].str.replace('K-A', 'E')

Посчитаем, сколько явных дубликатов в датасете.

In [131]:
games.duplicated().sum()

np.int64(241)

В датасете 241 дублирующихся строк. Удалим повторы.

In [132]:
# удалим дубликаты
games = games.drop_duplicates()

Проверим, сколько строк было удалено в процессе предобработки, чтобы убедиться, что они составляют не больше 10% от всех данных.

In [133]:
# посчитаем кол-во строк в датасете после предобработки 
rows_after = games.shape[0]

# найдем, сколько строк было удалено и какую долю они составляют
print(f'Было удалено {initial_len-rows_after} строк, это примерно {round((initial_len-rows_after)/initial_len,2)} всех строк.')

Было удалено 243 строк, это примерно 0.01 всех строк.


В ходе предобработки данных мы:
- привели названия столбцов к единому стилю snakecase
- удалили 2 строки с пропусками в поле `name`
- заменили пропуски в полях `eu_sales` и `jp_sales` на средние значения
- заменили пропуски в полях `critic_score` и `user_score` на -1
- заменили пропуски в поле `rating` на "нет данных"
- привели все `genre` к нижнему регистру, чтобы избавиться от дубликатов
- заменили устаревшее название `rating` "K-A" на "E", чтобы избавиться от дубликатов
- удалили явные дубликаты (их было 241)

В итоге было удалено 243 строки, примерно 0.01 от всех строк, что не исказит результаты анализа.

---

## 3. Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с `2000 по 2013` год включительно. Отберем данные по этому показателю. Сохраним новый срез данных в отдельном датафрейме `games_actual`.

In [141]:
# проведем фильтрацию по полю year_of_release
games_actual = pd.DataFrame(games[(games['year_of_release'] >= 2000) & (games['year_of_release'] <= 2013)])
# проверим правильность преобразования, выведем минимальный и максимальный год 
print(games_actual['year_of_release'].min())
games_actual['year_of_release'].max()

2000.0


2013.0

Мы получили датафрейм, в котором хранятся данные только за интересующие нас года.

---

## 4. Категоризация данных
    
Проведем категоризацию данных:
- Разделим все игры по оценкам пользователей и выделим такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [135]:
# создадим функцию для категоризации
def categorize(row):
    if row['user_score'] < 0:
        return 'Нет оценки'
    elif row['user_score'] < 3:
        return 'Низкая оценка'
    elif row['user_score'] < 8:
        return 'Средняя оценка'
    else:
        return 'Высокая оценка'
# создадим новый столбец 'user_score_category' с категорией оценки, применим функцию categorize к каждой строке датафрейма
games_actual['user_score_category'] = games_actual.apply(categorize, axis = 1)

- Разделим все игры по оценкам критиков и выделим такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [136]:
# создадим функцию для категоризации
def categorize(row):
    if row['critic_score'] < 0:
        return 'Нет оценки'
    elif row['critic_score'] < 30:
        return 'Низкая оценка'
    elif row['critic_score'] < 80:
        return 'Средняя оценка'
    else:
        return 'Высокая оценка'
# создадим новый столбец 'critic_score_category' с категорией оценки, применим функцию categorize к каждой строке датафрейма
games_actual['critic_score_category'] = games_actual.apply(categorize, axis = 1)

- После категоризации данных проверим результат: сгруппируем данные по выделенным категориям и посчитайте количество игр в каждой категории.

In [137]:
# сгруппируем данные по категории оценок пользователей и посчитаем сколько игр в каждой категории
user_scores = games_actual.groupby('user_score_category')['name'].count()
user_scores

user_score_category
Высокая оценка    2286
Нет оценки        6298
Низкая оценка      116
Средняя оценка    4081
Name: name, dtype: int64

Больше всего игр по оценкам пользователей в категории "нет оценки" и "средняя оценка", меньше всего в категории "низкая оценка".

In [138]:
# сгруппируем данные по категории оценок критиков и посчитаем сколько игр в каждой категории
critic_scores = games_actual.groupby('critic_score_category')['name'].count()
critic_scores

critic_score_category
Высокая оценка    1692
Нет оценки        5612
Низкая оценка       55
Средняя оценка    5422
Name: name, dtype: int64

Больше всего игр по оценкам критиков также в категории "нет оценки" и "средняя оценка", меньше всего в категории "низкая оценка".

- Выделим топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [139]:
# посчиатем сколько игр было выпущено на каждой платформе
top = games_actual.groupby('platform')['name'].count()
# отсортируем по убыванию и выведем первые 7 строк
top.sort_values(ascending = False).head(7)

platform
PS2     2127
DS      2120
Wii     1275
PSP     1180
X360    1121
PS3     1087
GBA      811
Name: name, dtype: int64

Больше всего игр за 2000-2013 гг. было выпущено на платформе `PS2` (2127 игр). Также близка по количеству выпущенных игр платформа `DS` (2120), на которой вышло на 7 игр меньше, чем на PS2.

---

## 5. Итоговый вывод


В ходе данного проекта мы проделали следующую работу:
- загрузили данные из датасета `new_games.csv`, познакомились с данными
- привели названия столбцов к единому стилю, чтобы было удобнее работать с ними
- преобразовали все столбцы к оптимальному типу данных
- посчитали количество пропусков в датасете, обработали их: удалили, заменили на индикаторы, заменили на средние значения
- убрали явные и неявные дубликаты
- провели фильтрацию: оставили срез данных только за 2000-2013 гг.
- добавили поля `user_score_category` и `critic_score_category`, в которых указана категория оценки: нет оценки, низкая, средняя, высокая; посчитали сколько игр в каждой категории 
- нашли топ-7 платформ по количеству игр, выпущенных за весь актуальный период